io/dagreader.go

```diff
@@ -66,12 +76,421 @@ func NewDagReader(ctx context.Context, n ipld.Node, serv ipld.NodeGetter) (DagRe
+// through the `dagWalker` in `io.SeekCurrent`.
+func (dr *dagReader) Seek(offset int64, whence int) (int64, error) {
+	switch whence {
+	case io.SeekStart:
+		if offset < 0 {
+			return -1, errors.New("invalid offset")
+		}
+
+		if offset == dr.offset {
+			return offset, nil
+			// Already at the requested `offset`, nothing to do.
+		}
+
+		left := offset
+		// Amount left to seek.
+
+		// Seek from the beginning of the DAG.
+		dr.resetPosition()
+
+		// Use the internal reader's context to fetch the child node promises
+		// (see `ipld.NavigableIPLDNode.FetchChild` for details).
+		dr.dagWalker.SetContext(dr.ctx)
+		// TODO: Performance: we could adjust here `preloadSize` of
+		// `ipld.NavigableIPLDNode` also, when seeking we only want
+		// to fetch one child at a time.
+
+		// Seek the DAG by calling the provided `Visitor` function on every
+		// node the `dagWalker` descends to while searching which can be
+		// either an internal or leaf node. In the internal node case, check
+		// the child node sizes and set the corresponding child index to go
+		// down to next. In the leaf case (last visit of the search), if there
+		// is still an amount `left` to seek do it inside the node's data
+		// saved in the `currentNodeData` buffer, leaving it ready for a `Read`
+		// call.
+		err := dr.dagWalker.Seek(func(visitedNode ipld.NavigableNode) error {
+			node := ipld.ExtractIPLDNode(visitedNode)
+
+			if len(node.Links()) > 0 {
+				// Internal node, should be a `mdag.ProtoNode` containing a
+				// `unixfs.FSNode` (see the `balanced` package for more details).
+				fsNode, err := unixfs.ExtractFSNode(node)
+				if err != nil {
+					return err
+				}
+
+				// If there aren't enough size hints don't seek
+				// (see the `io.EOF` handling error comment below).
+				if fsNode.NumChildren() != len(node.Links()) {
+					return io.EOF
+				}
+
+				// Internal nodes have no data, so just iterate through the
+				// sizes of its children (advancing the child index of the
+				// `dagWalker`) to find where we need to go down to next in
+				// the search.
+				for {
+					childSize := fsNode.BlockSize(int(dr.dagWalker.ActiveChildIndex()))
+
+					if childSize > uint64(left) {
+						// This child's data contains the position requested
+						// in `offset`, go down this child.
+						return nil
+					}
+
+					// Else, skip this child.
+					left -= int64(childSize)
+					err := dr.dagWalker.NextChild()
+					if err == ipld.ErrNextNoChild {
+						// No more child nodes available, nothing to do,
+						// the `Seek` will stop on its own.
+						return nil
+					} else if err != nil {
+						return err
+						// Pass along any other errors (that may in future
+						// implementations be returned by `Next`) to stop
+						// the search.
+					}
+				}
+
+			} else {
+				// Leaf node, seek inside its data.
+				err := dr.saveNodeData(node)
+				if err != nil {
+					return err
+				}
+
+				_, err = dr.currentNodeData.Seek(left, io.SeekStart)
```

Comment:
We should probably sanity check to make sure this worked (i.e., we seeked the right amount).